common/src/api/external/http_pagination.rs

```diff
@@ -298,15 +296,45 @@ impl ScanParams for ScanById {
 // We include this now primarily to exercise the interface for doing so.
 
 /// Query parameters for pagination by name or id
-pub type PaginatedByNameOrId =
-    PaginationParams<ScanByNameOrId, PageSelectorByNameOrId>;
+pub type PaginatedByNameOrId<Selector> = PaginationParams<
+    ScanByNameOrId<Selector>,
+    PageSelectorByNameOrId<Selector>,
```

Comment:
So in general, it seems like the first parameter to PaginationParams is going to want to be the tuple of NameOrIds needed to find the item in question. The second parameter is going to want to be just the ID of the top level container. That is to say, I think we should only be doing name lookups for the first query; subsequent pages should just have the ID of the container that we're iterating through.

Reply:
I tried this, but it's rather difficult and made the handlers a lot messier. 

Right now we take the query parameters and resolve them to a lookup and pass that to nexus. My understanding of this approach is that we'll need to resolve the lookup to an ID inside the handler as well as passing it along to be further resolved by Nexus (on the first page). That also means we have to have logic to check if it's a selector or id (or if the selector is an id) so we're not doing wasteful lookups. My attempts at this made both the handler much longer and harder to understand as well as increased the complexity of the types around pagination.

I agree that this would be an ideal approach. I also think it's an optimization that we can defer. From a user documentation perspective the same caveats would apply when looking up a resource by name to paginate as would when just looking up a name in general.

Reply:
> I tried this, but it's rather difficult and made the handlers a lot messier.

That's disappointing; I could imagine it being a way to extract common patterns rather than to complicate each handler.

> Right now we take the query parameters and resolve them to a lookup and pass that to nexus. My understanding of this approach is that we'll need to resolve the lookup to an ID inside the handler as well as passing it along to be further resolved by Nexus (on the first page).

Not necessarily. It could be as simple as having the nexus function that performs the resolution return the resolved ID.

> That also means we have to have logic to check if it's a selector or id (or if the selector is an id) so we're not doing wasteful lookups. My attempts at this made both the handler much longer and harder to understand as well as increased the complexity of the types around pagination.

It could literally be the same types if, for some reason, that's simpler: it could just be that for the page selector **only** the simplest ID form is encoded in the next_token.

> I agree that this would be an ideal approach. I also think it's an optimization that we can defer.

I think of this not as a performance optimization, but a simplification of the handlers. If it's not, agreed, we can defer or just not pursue it.

> From a user documentation perspective the same caveats would apply when looking up a resource by name to paginate as would when just looking up a name in general.

Further caveats apply: with this approach a rename of any intermediate collection will cause the iteration to fail whereas using a resolved ID for the container means that renames (e.g. of the parent) won't interfere.

Reply:
> Not necessarily. It could be as simple as having the nexus function that performs the resolution return the resolved ID.

Yeah, I was thinking the same. We turn the selector into a lookup and pass that along to the function that lists the resources. I tried returning a tuple of the resource list and ID but I had some issues with the return type no longer being a result list. Maybe it just needed to be a result of a tuple where one member was a result list? Uncertain.